src/main/java/org/elasticsearch/plugins/repository/LocalPluginRepository.java

```diff
@@ -0,0 +1,433 @@
+                        && StringUtils.equalsIgnoreCase(name, plugin.name())
+                        && StringUtils.equalsIgnoreCase(version, plugin.version());
+            }
+        });
+
+        // The same plugin can't exist multiple times... if so, reports the error
+        // because a manual clean up will be necessary
+        if ((plugins != null) && (!plugins.isEmpty())) {
+            if (plugins.size() > 1) {
+                throw new PluginRepositoryException(name(), "Plugin [organisation=" + organisation
+                        + ", name=" + name + ", version=" + version + "] exists in multiple instances");
+            }
+
+            return plugins.iterator().next();
+        }
+        return null;
+    }
+
+    @Override
+    public Collection<PluginDescriptor> list() {
+        return list(new Filter() {
+            @Override
+            public boolean accept(PluginDescriptor plugin) {
+                // Accepts all plugins
+                return true;
+            }
+        });
+    }
+
+    @Override
+    public Collection<PluginDescriptor> list(final Filter filter) {
+        // Get the es-plugin.properties filename
+        String esPluginProps = PluginsService.ES_PLUGIN_PROPERTIES;
+        if (settings != null) {
+            esPluginProps = settings.get(PluginsService.ES_PLUGIN_PROPERTIES_FILE_KEY, PluginsService.ES_PLUGIN_PROPERTIES);
+        }
+
+        final ImmutableList.Builder<PluginDescriptor> plugins = ImmutableList.builder();
+
+        String pluginName;
+        PluginDescriptor.Builder descriptor;
+
+        if (Files.exists(pluginsDir)) {
+            try (DirectoryStream<Path> root = Files.newDirectoryStream(pluginsDir)) {
+                for (Path pluginDir : root) {
+                    if (Files.isDirectory(pluginDir)) {
+
+                        // Determine the name of the plugin from the directory's name
+                        pluginName = pluginDir.getFileName().toString();
+                        descriptor = pluginDescriptor(pluginName);
+
+                        // Introspect JAR files to find plugin's description
+                        try (DirectoryStream<Path> jarFiles = Files.newDirectoryStream(pluginDir, "*" + JAR)) {
+                            for (Path jarFile : jarFiles) {
+                                // Indicates that the plugin has code/library
+                                descriptor.jvm();
+
+                                String fileName = jarFile.getFileName().toString();
+                                if (!fileName.contains(pluginName)) {
+                                    continue;
+                                }
+
+                                // Constructs the plugin's descriptor from the JAR filename...
+                                // Since plugin's JAR file can have various filenames,
+                                // the result is incertain
+
+                                // Removes "elasticsearch-" prefix
```

Comment:
It took me some seconds to understand what was the goal of the 4 next lines... May be a general comment and grouping the 4 lines?

```java
// We try to get the plugin version by subtracting `elasticsearch-` prefix, `.jar` suffix, the plugin name itself.
// If any, we should have at the end `-version`
String s = StringUtils.removeStart(fileName, "elasticsearch-");
s = StringUtils.removeEnd(s, JAR);
s = StringUtils.removeStart(s, pluginName);
String version = StringUtils.substringAfterLast(s, "-");
if (Strings.hasText(version)) {
    descriptor.version(version);
}
```